### DISCUSSION TITLE
Clean up user handling docmentation

### DISCUSSION THREAD

**Discussion by trickkiste** [ltsp/ltsp#294] [Q&A]
Hi Alkis,
I noticed that LTSP documentation lacks in clarity when it comes to user handling.
I guess the underlying system has changed much the last year and the documentation still shows bits and pieces of different stages. 

I tried to compile a picture using these documents, but many questions remain:
https://github.com/ltsp/ltsp/blob/d34a142b22e08228de71324ca1d7f022a5031bd2/docs/ltsp.conf.5.md
https://github.com/ltsp/ltsp/blob/046af079fc85a0c6e0bd90c2bf1222265f0f9505/ltsp/client/login/pwmerge
https://github.com/ltsp/ltsp/blob/master/ltsp/client/login/pamltsp.md

For one, the documentation does not clarify at all, what actually is the source for sur and sgr and what is the destination. Both scenarios seem plausible, the source being my build image or the server. 

Also there is PASSWORDS_x in ltsp.conf and then there is pamltsp entries in shadow.

I also could not figure out, how I can use the mergepw mechanism to merge user accounts from my server system via ltsp.img (initrd) into my images passwd and shadow, while perserving local auth.

Please forgive me, that this report is not well structured and conclusive, I assembled these bits last before going to bed.

Is there a document available which I might have overlooked, explaining all that?

Where possible I would love to help cleaning things up, but currently I lack the proper understanding to do so.

Cheers and thank you for all the great work on LTSP so far.
Markus

**Comment by alkisg** [comment]
Hi Markus,
how do you plan to mount /home for your users?

**Comment by trickkiste** [comment]
Hi Alkis,

/home is mounted via CephFS in my setup (CephFS is a POSIX compliant
distributed network filesystem).
And my boot device is served via RBD (Rados Block Device, a distributed
network block device, which I dd the squashfs image to).
Both things live in my Ceph storage cluster, so I only have one
distributed, networked, bottleneck-free storage cluster to maintain.
There would also be an S3/SWIFT compatible object storage layer, which I am
currently not using. Maybe a place to store authentication info in the
future.

So, I mount the complete /home dir to the clients with all home dirs in it.

Therefore all I need would be the accounts in ltsp.img (initrd) being
merged without modifications. And for that I would like to select a certain
usergroup and copy only the accounts belonging to this group (as in having
this group defined as their primary group in passwd [not being listed as
group members in groups]. Not sure if pwmerge does support that currently,
have not yet studied all the code]). Is that currently supported?

Mit besten Grüßen,
Markus Kienast

--

Markus Kienast - CEO

FLANEUR TV
<mark@sfe.tv>mark@f <mark@trickkiste.at>laneur.tv
+43 699 18084401

Commercial Register: FN 363780 x
UID: ATU66601545
WWW: http:// <http://trickkiste.at/>flaneur.tv


Am Mo., 6. Juli 2020 um 04:39 Uhr schrieb Alkis Georgopoulos <
notifications@github.com>:

**Comment by trickkiste** [comment]
Ceph actually provides two authentication mechanisms:
* One for CephFS and RBD based on an access key.
* One for S3/SWIFT RadosGW, which works the same way as Amazon S3.
Maybe one could hack these to provide PAM authentication for users. Thereby
the whole user authentication infrastructure would at the same time become
distributed, replicated and therefore redundantly available.

CephFS allows for per directory read/write permissions on the FS level
additionally to the standard unix permissions. Could come in handy!
Also I am currently thinking about serving the boot images via CephFS as
well in the future instead of going the RBD way.
Simplifies the setup and I don't need to "dd" the images anymore. They
could be directly generated into the CephFS tree on the server.

Sound good?

Mit besten Grüßen,
Markus Kienast

--

Markus Kienast - CEO

FLANEUR TV
<mark@sfe.tv>mark@f <mark@trickkiste.at>laneur.tv
+43 699 18084401

Commercial Register: FN 363780 x
UID: ATU66601545
WWW: http:// <http://trickkiste.at/>flaneur.tv


Am Mo., 6. Juli 2020 um 18:17 Uhr schrieb Markus Kienast <mark@flaneur.tv>:

**Comment by alkisg** [comment]
Hi Markus,

please do not send all the old email content in the github replies, as one can just [scroll up the page](https://github.com/ltsp/community/issues/143#issuecomment-654341079) and read them.

From what I understand, what you're looking for is the `ltsp image --cleanup=0` parameter. This will keep all shadow entries in your exported image (it's documented in its man page).

Note that this is disabled by default for a good reason. **We do not want to make password hashes (shadow files) public.**

If you put them in ltsp.img, anyone can get them via tftp. If you put them in CephFS or NFS, anyone can get them from there. Then, they can run brute force algorithms and possibly crack them.

This is why we want user accounts and shadow files to reside in the server, either /etc/shadow or even /not/exported/path/chroot/etc/shadow. That way, users can try to authenticate via SSH, which will authenticate against the password hash without exposing it to the local network.

So I would advice against making password hashes public.

**Comment by trickkiste** [comment]
Hi Alkis!
Thank you for your quick response.

OK, I understand the shadow problem. I fully agree.

The documentation hints that I can do a pure SSH auth without SSHFS too.
How do I achieve this?

ltsp image --cleanup 0 is only partly useful for me.

Because I do not generate my boot images from the root filesystem but much
rather have a separate RBD disk, which I spin up connected to a VM or in RW
mode from a client, when I install stuff. So my users are not present there
and on the other hand, I want to pass them on not via the boot image but
the initrd ltsp.img, so I do not have to regenerate the image on every user
I add.

Here is the process, I would like to perform:
1. generate my boot image from the mounted RBD fs. (No problem)
2. add whichever new user accounts I need to my LTSP server system while
assigning them a special "staff" group as their primary group. (no problem)
3. generate an initrd (ltsp.img) holding the user information from my LTSP
server system (not from the VM). DONE, this is the standard behaviour.
4. configure ltsp.conf in a way, that all users having primary group
"staff" will be added to the clients passwd file and set for SSH
authentication (but not SSHFS mount, as my home is already mounted).
5. authenticate via SSH against the LTSP server.

Is there a way to do this?
Also I would like to auto login some accounts.

Cheers,
Markus

**Comment by alkisg** [comment]
All the 5 steps that you mention are the default behaviour.
SSH is used for authentication is /home is already mounted, which in your case, is it.
If /home was not a mountpoint, SSHFS would be used for both authentication and /home/username mounting.

To configure autologins and passwordless logins, read the ltsp.conf man page, you don't need to tweak anything at all.

I.e. you just need to make sure that /home is mounted via FSTAB_x, everything else should work fine with the default setup.

**Comment by trickkiste** [comment]
Thank you Alkis for the clarification.
These things are not clear from the documentation though. If I find the time, I happily submit a documentation fix for this.

But what is the central place for documentation now.
I wanted to point out, that currently I could not find sufficient info to develop the understanding you just provided.

There is doc in the git repo, in man pages, on the ltsp.org website and the most extensive one in the code itself. While all these sources still did not provide a clear picture fast and effortlessly.

Please understand, this is NOT me complaining. I am just giving feedback and pointing out things. And the purpose is not even to say "please fix this all yourself" but much rather to say "please clarify these things and your documentation concept, so I can contribute some patches in a way, that fits your documentation concept".

BTW, there still seems to be some bug in pwmerge. I'll post the report in a bit.

Which makes step Nr. 3 in my scenario to not work. 

Anyhow, I'll put everything in the boot image for now and will hunt down this bug in pwmerge later. Also I believe a way to select by UID/GID numgers and ranges should be added. I believe this might be much more practical for most use cases.

Thanks again,
Markus

PS: The divide between ltsp/ltsp and ltsp/community just confused the heck out of me. I was looking for this bug report and did not find it anymore, later realizing, I filed it against ltsp/community.

**Comment by alkisg** [comment]
> PS: The divide between ltsp/ltsp and ltsp/community just confused the heck out of me. 

You filed this issue in ltsp/ltsp, but then I thought it wasn't a documentation issue, so I moved it to community. But if you're planning to send specific documentation patches, we could move it back there.

The documentation goal is:

- Whatever can go to the man pages, should go there (and in ltsp.org/man too of course)
- The rest should be in ltsp.org/docs etc
- And the community can maintain the community wiki.

Pamltsp needs to be rewritten as an `ltsp login` applet before more documentation is written about it.

**Comment by trickkiste** [comment]
One more question:
How does pamltsp know, which server to connect to for authentication?
I can't get this to work properly.

I have verified that my user accounts are properly merged, using a local account to check these files on the client. I can also SSH into the server machine, which has the IP 10.101.0.10 with the accounts in question successfully. However, attempting to login via display manager does fail every time with: "Login incorrect".

However, I can't see the GUI login attempts in the servers auth.log, so it seems, pamltsp is trying to connect to the wrong server.

DHCP is done by MAAS in my RBD/CephFS setup.

Also I want to point out, that pwmerge documentation does not clarify what SUR SGR DUR DGR actually means in the LTSP setup. Yes, source_dir destination_dir merge_dir are somewhat clear, but what do you consider source in terms of the LTSP workflow.

It should be mentioned that "source" is the user DB of the system you run "sudo ltsp initrd" on and "destination" is whatever lives in the boot image under /etc and the merge dir will also be an overlay dir on /etc on the client machine.

That information missing can lead to quite some confusion.

**Comment by alkisg** [comment]
pamltsp tries to connect to the server named "server".
LTSP puts an entry for server in /etc/hosts, and it sets up the correct SSH known hosts for it.
All this works fine by default, there's no configuration necessary.
It would help if you described what you're trying to do. E.g. if you're using another server than the LTSP server, you'd need to provide the SSH known hosts for it.

You can try to login from vt1 (Ctrl+Alt+F1) to see all the PAM error messages, as most display managers only show the last error.

**Comment by trickkiste** [comment]
Thank you Alkis,

for some reason generation of the automatic /etc/hosts entry did not work: it showed only "server" but no IP.
Putting the SERVER config option in ltsp.conf did fix the problem though.

My boot cmdline looks like this:
vmlinuz initrd=ltsp.img initrd=initrd.img boot=rbd ro rbdroot=10.101.0.25,10.101.0.27,10.101.0.28:ltsp-boot:XXXXXXXXXXXXXXXXXXXXX==:squashfs:ltsp-01:: pcie_aspm=off pcie_port_pm=off BOOTIF=01-1c-69-7a-0c-80-71

Please notice, that the IPs mentioned under "rbdroot" are pure Ceph storage nodes not the SSH server. So extracting the IP from the command line would not work in this case.

For the same reason "ss -tn" also did not provide the information needed for the "sed" operation in your code:
```
    if [ -n "$cmd" ]; then
        SERVER=$(rw $cmd |
            sed -n 's/.*[[:space:]]\([0-9.:]*\):\([0-9]*\)[^0-9]*/\1/p' |
            head -1)
    fi
```

But apparently there were enough space characters for this section not to kick in:
```
    # If unable to detect, default to those:
    SERVER=${SERVER:-$IPCONFIG_ROOTSERVER}
    SERVER=${SERVER:-$GATEWAY}
    SERVER=${SERVER:-192.168.67.1}
```

When I finally set the SERVER var in ltsp.conf, I noticed that I get an error/info message on login, if the "server" does not have the users home directories mounted. It did login successfully however and yes, one usually would have home mounted of course. But currently I have it mounted in another spot.

Thanks again, for your help.

All in all, everything is working quite nicely now and the RBD/CephFS setup is a treat!
In a couple of month I will try again to submit a useful patch to provide a universal method for allowing alternative boot options like mine to work out of the box. Hope we can agree on a solution that both of us find an acceptable compromise.
